storage/lmdb/src/error.rs

```diff
@@ -0,0 +1,91 @@
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use lmdb::Error;
+use std::io::{Error as IoError, ErrorKind};
+
+use storage_core::error::{Fatal, Recoverable};
+
+fn process<T>(err: Error, default: storage_core::Result<T>) -> storage_core::Result<T> {
+    match err {
+        // Not found is signified by succeeding with a value signifying a missing entry
+        Error::NotFound => default,
+
+        // Overwriting a key is not an error
+        Error::KeyExist => default,
+
+        // Initialization issues
+        Error::Invalid | Error::VersionMismatch => Err(Recoverable::DbInit.into()),
+
+        // Transaction failed to commit
+        Error::BadTxn => Err(Recoverable::TransactionFailed.into()),
+
+        // We have exhausted some resource which may become available again later
+        Error::ReadersFull | Error::TxnFull | Error::MapFull | Error::TlsFull => {
+            Err(Recoverable::TemporarilyUnavailable.into())
+        }
+
+        // These signify an implementation flaw
+        Error::BadDbi | Error::Panic | Error::CursorFull | Error::PageFull | Error::BadRslot => {
+            Err(Fatal::InternalError.into())
+        }
+
+        // These signify the database flags are not in sync with the schema
+        Error::DbsFull | Error::BadValSize | Error::Incompatible => {
+            Err(Fatal::SchemaMismatch.into())
+        }
+
+        // These are database corruption issues
+        Error::PageNotFound | Error::Corrupted => Err(Fatal::DatabaseCorrupted.into()),
+
+        // Other errors
+        Error::Other(errno) => {
+            let err = IoError::from_raw_os_error(errno);
+
+            // Try to recover the error from the system to get a more informative message
+            let last_err = IoError::last_os_error();
+            let same_err = err.kind() == last_err.kind();
+            let err = if same_err { last_err } else { err };
+
+            // Classify recoverable vs. fatal I/O errors
+            Err(match err.kind() {
+                ErrorKind::BrokenPipe
+                | ErrorKind::AlreadyExists
+                | ErrorKind::PermissionDenied
+                | ErrorKind::NotFound => Recoverable::Io(err.kind(), err.to_string()).into(),
+                _ => Fatal::Io(err.kind(), err.to_string()).into(),
+            })
+        }
+
+        // Make this an error for the time of being.
+        // TODO: Would be nice to handle memory map resizes automatically somehow but it looks hard.
```

Comment:
Monero does that. You can see how they do it. It's not really that hard: https://github.com/monero-project/monero/blob/fc907a957078cae2dc68348886a33363848dd089/src/blockchain_db/lmdb/db_lmdb.cpp#L494

On the other hand, given that we have recoverable errors, and assuming that this error happens only for transactions, we could just retry.

Reply:
Filed an issue for this #483